### PR TITLE
Add 404 Not Found

### PR DESCRIPTION
**Pull request by christophermanahan**
### Description

- Add a Router that can be configured by the client
- Add a 404 Not Found Response that will be sent if a uri has not been added to the router
- Add a Route class that associates routes with handler function

### CODE REVIEW

File: src/main/java/io/github/christophermanahan/carnitas/RequestParser.java [hunk 1]
```diff
@@ -62,9 +73,11 @@ private boolean isContentLength(Optional<String> header) {
...
-    private Optional<HTTPRequest> request(Optional<String> method, Optional<String> body) {
-        return method
-          .map(HTTPRequest::new)
-          .map(request -> request.withBody(body));
+    private Optional<HTTPRequest> request(Optional<String> method, Optional<String> uri, Optional<String> body) {
+        if (method.isPresent() && uri.isPresent()) {
+            return Optional.of(new HTTPRequest(method.get(), uri.get()).withBody(body));
+        } else {
+            return Optional.empty();
+        }
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:81]: We talked about what we could to do clean this up.

File: src/main/java/io/github/christophermanahan/carnitas/RequestParser.java [hunk 2]
```diff
@@ -5,25 +5,36 @@
 
 public class RequestParser implements Parser {
     public Optional<HTTPRequest> parse(Reader reader) {
-        Optional<String> method = getMethod(reader);
+        Optional<String> statusLine = readLine(reader);
+        Optional<String> method = getMethod(statusLine);
+        Optional<String> uri = getUri(statusLine);
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:10]: We talked about adding a request line abstraction maybe?

File: src/main/java/io/github/christophermanahan/carnitas/Main.java [hunk 3]
```diff
@@ -27,4 +27,13 @@ private static int port(String[] args) {
           .map(Integer::parseInt)
           .orElse(33333);
     }
+
+    private static Router router() {
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Main.java]: We talked about making Routers just a different type of Handler

**Comment by avni510** [comment]
I think you'll want to be careful with the distinction of a `Handler` and `Router` they both have 2 different responsibilities and purposes. If you need to bounce ideas off someone or want to talk more about this I'm happy to chat

**Review by damonkelley — COMMENTED** [review]
I forgot to post these comments yesterday! Sorry about that!

File: src/main/java/io/github/christophermanahan/carnitas/Main.java [hunk 4]
```diff
@@ -27,4 +27,13 @@ private static int port(String[] args) {
           .map(Integer::parseInt)
           .orElse(33333);
     }
+
+    private static Handler router() {
```
Comment (avni510) [inline src/main/java/io/github/christophermanahan/carnitas/Main.java]: I think this should pulled out into a `Configuration` class

Comment (damonkelley) [reply src/main/java/io/github/christophermanahan/carnitas/Main.java]: Yeah, what is this `router`? What is its purpose? It seems to me that it is the application that will help satisfy our acceptance tests.

File: src/main/java/io/github/christophermanahan/carnitas/Main.java (hunk 3 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Main.java]: This is from an outdated commit and has been fixed

File: src/main/java/io/github/christophermanahan/carnitas/RequestParser.java (hunk 1 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:81]: I will be working on this with any time leftover after completing this weeks iteration

File: src/main/java/io/github/christophermanahan/carnitas/RequestParser.java (hunk 2 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/RequestParser.java:10]: I will also be working on this after completing this weeks iteration!

File: src/main/java/io/github/christophermanahan/carnitas/Main.java (hunk 4 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Main.java]: It has been pulled out into an application class that provides a router to satisfy the acceptance tests.

File: src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java [hunk 5]
```diff
@@ -3,18 +3,31 @@
...
-        this.statusCode = statusCode;
+    enum StatusCode {
+        OK ("200 OK"),
+        CREATED("201 Created"),
+        NOT_FOUND("404 Not Found");
+
+        final String value;
+
+        StatusCode(String value) {
+            this.value = value;
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java]: Is `value` in the domain of status codes? Is there a better name for this?

File: src/main/java/io/github/christophermanahan/carnitas/Main.java [hunk 6]
```diff
@@ -9,7 +9,7 @@ public static void main(String[] args) {
         try (ServerSocket serverSocket = new ServerSocket(port(args))) {
             new HTTPServer(
               new RequestParser(),
-              new RequestHandler(),
+              Application.router(),
```
Comment (damonkelley) [inline src/main/java/io/github/christophermanahan/carnitas/Main.java]: I don't love the API of the `Application`, but I suppose it is fine for now. I would prefer to see something that looks like.
```suggestion
              new Application(),
```

File: src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java (hunk 5 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/HTTPResponse.java]: @damonkelley Updated to Status.code

File: src/main/java/io/github/christophermanahan/carnitas/Main.java (hunk 6 above)
Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Main.java]: I also agree that this would be an improvement, I will keep this in mind.

Comment (christophermanahan) [reply src/main/java/io/github/christophermanahan/carnitas/Main.java]: API improved!

**Review by damonkelley — APPROVED** [review]

**Pull request merged**